scripts/question_answering/train.py

```diff
@@ -0,0 +1,227 @@
+        end = nd.array([x[5] for x in batch_data])
+        batch_sizes = context.shape[0]
+        context = gluon.utils.split_and_load(
+            data=context,
+            ctx_list=CTX
+        )
+
+        query = gluon.utils.split_and_load(
+            data=query,
+            ctx_list=CTX
+        )
+        context_char = gluon.utils.split_and_load(
+            data=context_char,
+            ctx_list=CTX
+        )
+        query_char = gluon.utils.split_and_load(
+            data=query_char,
+            ctx_list=CTX
+        )
+        begin = gluon.utils.split_and_load(
+            data=begin,
+            ctx_list=CTX
+        )
+        end = gluon.utils.split_and_load(
+            data=end,
+            ctx_list=CTX
+        )
+
+        with autograd.record():
+            different_ctx_loss = [loss_function(*model(c, q, cc, qc, b, e))
+                                  for c, q, cc, qc, b, e in
+                                  zip(context, query, context_char, query_char, begin, end)]
+
+            for loss in different_ctx_loss:
+                loss.backward()
+        if global_step == 1:
+            for name, param in model.collect_params().items():
+                ema.add(name, param.data(CTX[0]))
+        trainer.set_learning_rate(warm_up_lr(global_step))
+        trainer.allreduce_grads()
+        reset_embedding_grad(model)
+        tmp = []
+        for name, paramater in model.collect_params().items():
+            grad = paramater.grad(context[0].context)
+            if name == 'qanet0_embedding0_weight':
+                grad[0:2] += opt.weight_decay * \
+                    paramater.data(context[0].context)[0:2]
+            else:
+                grad += opt.weight_decay * paramater.data(context[0].context)
+            tmp.append(grad)
+        gluon.utils.clip_global_norm(tmp, opt.clip_gradient)
+        reset_embedding_grad(model)
+        trainer.update(batch_sizes, ignore_stale_grad=True)
+        for name, param in model.collect_params().items():
+            ema(name, param.data(CTX[0]))
+
+        batch_loss = .0
+        for loss in different_ctx_loss:
+            batch_loss += loss.mean().asscalar()
+        batch_loss /= len(different_ctx_loss)
+        total_loss += batch_loss
+
+        batch_train_ce.append([global_step, batch_loss])
+        accum_avg_train_ce.append([global_step, total_loss / step])
+
+        print('batch %d/%d, total_loss %.2f, batch_loss %.2f' %
+              (step, total_batchs, total_loss / step, batch_loss), end='\r', flush=True)
+        nd.waitall()
+
+
+def initial_model_parameters(model):
+    r"""
+    Initial the word embedding layer.
+    """
+    model.collect_params().initialize(ctx=CTX)
+    # initial embedding parameters with glove
+    word_embedding = nd.array(load_emb_mat(opt.word_emb_file_name))
+
+    model.word_emb[0].weight.set_data(word_embedding)
+
+
+def reset_embedding_grad(model):
+    r"""
+    Reset the grad about word embedding layer.

```

Comment:
Does it mean you want to fix the embedding layer? If so, please disable gradients for the word embeddings instead of manually setting them to zero. That will save both GPU memory and computation